_source/_posts/2021-08-02-fix-common-problems-cors.md

```diff
@@ -0,0 +1,530 @@
+
+```go
+package main
+
+import (
+	"github.com/gin-contrib/static"
+	"github.com/gin-gonic/gin"
+)
+
+func main() {
+	r := gin.Default()
+	r.Use(static.Serve("/", static.LocalFile("./frontend", false)))
+	r.Run(":8080")
+}
+```
+
+This code simply serves the contents of the `frontend` directory on requests on port 8080. Note that JavaScript makes a call to port `http://localhost:8000` which is a separate service.
+
+Start the server and point a web browser at `http://localhost:8080` to see the static content.
+
+```bash
+go run frontend.go
+```
+## How to Build a Simple REST API in Go
+
+Create a directory called `rest` to contain the code for a basic REST API.
+
+**NOTE:** A separate directory is required as Go doesn't allow two program entry points in the same directory.
+
+Create a file called `rest/server.go` containing the following Go code:
+
+```go
+package main
+
+import (
+	"fmt"
+	"strconv"
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+)
+
+var messages []string
+
+func GetMessages(c *gin.Context) {
+	version := c.Param("version")
+	fmt.Println("Version", version)
+	c.JSON(http.StatusOK, gin.H{"messages": messages})
+}
+
+func main() {
+	messages = append(messages, "Hello CORS!")
+	r := gin.Default()
+	r.GET("/api/:version/messages", GetMessages)
+	r.Run(":8000")
+}
+```
+
+A list called `messages` is created to hold message objects.
+
+The function `GetMessages()` is called whenever a GET request is made to the specified URL. It returns a JSON string containing the messages. The URL contains a path parameter which will be `v1` or `v2`. The server listens on port 8000.
+
+The server can be run using:
+
+```bash
+go run rest/server.go
+```
+
+## How to Solve a Simple CORS Issue
+
+We now have two servers—the front-end one on `http://localhost:8080`, and the REST API server on `http://localhost:8000`. Even though the two servers have the same hostname, the fact that they are listening on different port numbers puts them in different domains from the CORS perspective. The domain of the web content is referred to as the origin. If the JavaScript `fetch` request specifies `cors` a request header will be added identifying the origin.
+
+```http
+Origin: http://localhost:8080
+```
+
+Make sure both the frontend and REST servers are running.
+
+Next, point a web browser at `http://localhost:8080` to display the web page. We are going to get JavaScript errors, so open your browser's developer console so that we can see what is going on. In Chrome it is *View** > **Developer** > **Developer Tools**.
+
+Next, click on the **Send v1** button. You will get a JavaScript error displayed in the console:
+
+```
+Access to fetch at 'http://localhost:8000/api/v1/messages' from origin 'http://localhost:8080' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
+```
```

Comment:
This message doesn't wrap in the final output. I believe if you add "bash", it might. Not sure.

Reply:
Good catch!
`bash` didn't work, but I'm switching them to a quote block.  I figured out how to do this with asciidoc recently, but not with markdown